### PR TITLE
feat: scaling with reschedule semantics

### PR DESCRIPTION
**Pull request by shanicky**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR introduces a scaling mechanism for the reschedule semantics

Support for scale in/out has been added through modifications to the original migration framework
This PR uses a simple rebalance function to evenly distribute the virtual nodes after scaling.
The logic for scaling and migration will be merged in a later update (which is still a special case now).

Of course, scaling supports cascading updates

Note that this PR is only a modification of the migration framework, for the fragment that cannot migrate also cannot scale

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
https://github.com/risingwavelabs/risingwave/issues/3750

### CODE REVIEW

File: src/meta/src/stream/scale.rs [hunk 1]
```diff
@@ -672,16 +922,36 @@ where
                 None
             };
 
+            let mut vnode_bitmap_updates = fragment_updated_bitmap.remove(&fragment_id).unwrap();
+
+            for actor_id in actors_after_reschedule.keys() {
```
Comment (BugenZhao) [inline src/meta/src/stream/scale.rs:927]: May add some comments here: only keep the changed bitmaps.

File: src/meta/src/stream/scale.rs [hunk 2]
```diff
@@ -897,11 +1169,225 @@ where
...
+
+                            for (actor_id, bitmap) in downstream_updated_bitmap {
+                                for (idx, pos) in raw.iter_mut().enumerate() {
+                                    if bitmap.is_set(idx).unwrap() {
+                                        *pos = *actor_id;
+                                    }
+                                }
+                            }
+
+                            let (original_indices, data) = compress_data(&raw);
```
Comment (BugenZhao) [inline src/meta/src/stream/scale.rs:1187]: Looks like a little bit duplicated with below. Can we extract it? (Or we may introduce some specific structs for mapping, bitmap stuffs in the future)
https://github.com/risingwavelabs/risingwave/blob/752dd2b70fbf48ac8a1cb24379a4fa0a07aff4cd/src/meta/src/stream/scale.rs#L870-L892

**Review by BugenZhao — COMMENTED** [review]
The code looks so successful to me! I'll further test the functionality with cli tools.

File: src/meta/src/manager/catalog/fragment.rs [hunk 3]
```diff
@@ -629,6 +604,65 @@ where
...
+                            .into_iter()
+                            .zip_eq(added_parallel_units.into_iter())
+                            .collect();
+
+                        for parallel_unit_id in &mut vnode_mapping.data {
+                            if let Some(target) = replace_map.get(parallel_unit_id) {
+                                *parallel_unit_id = *target;
+                            }
+                        }
+                    } else {
```
[the remark this answers is not in the document]
Comment (shanicky) [reply src/meta/src/manager/catalog/fragment.rs:619]: Yes, but maybe more tests are needed to cover the corner case

File: src/meta/src/stream/scale.rs [hunk 4]
```diff
@@ -90,6 +96,182 @@ impl RescheduleContext {
...
+        .map(|actor_id| Balance {
+            actor_id: *actor_id,
+            balance: -(expected as i32),
+            builder: BitmapBuilder::zeroed(VIRTUAL_NODE_COUNT),
+        })
+        .collect_vec();
+
+    for balance in created_balances
+        .iter_mut()
+        .rev()
```
[the remark this answers is not in the document]
Comment (shanicky) [reply src/meta/src/stream/scale.rs:197]: The front of this queue is the removed actor, queued in descending order of balance, and the tail is the created actor s, which needs to be assigned the extra 1 from the tail to these created actors, so that in the case of migration,there will not be 1 remaining vnode of the deleted node.

For example, when we have 5 actors assigned 256 vnodes, there will definitely be 1 actor with 52 vnodes (the rest are 51), When we do a -1+1 migration, since the average vnode per actor is still 51 after the migration, we need to assign that extra 1 to the migrated new actor, otherwise there will be a situation where 51 are assigned to the new actor and the remaining 1 to the remain old actor, which defeats the purpose of migration

File: src/stream/src/executor/merge.rs [hunk 5]
```diff
@@ -132,32 +132,35 @@ impl MergeExecutor {
...
+                                    )
+                                })
+                                .try_collect()
+                                .map_err(|e| anyhow!("failed to create upstream receivers: {e}"))?;
+
+                            // Poll the first barrier from the new upstreams. It must be the same as
+                            // the one we polled from original
+                            // upstreams.
+                            let mut select_new =
+                                SelectReceivers::new(self.actor_context.id, new_upstreams);
```
[the remark this answers is not in the document]
Comment (shanicky) [reply src/stream/src/executor/merge.rs:157]: added

File: src/meta/src/manager/catalog/fragment.rs [hunk 6]
```diff
@@ -629,6 +604,65 @@ where
...
+                                .map(|actor_id| {
+                                    *actor_to_parallel_unit.get(actor_id).unwrap() as ParallelUnitId
+                                })
+                                .collect_vec();
+
+                            *vnode_mapping = ParallelUnitMapping {
+                                fragment_id,
+                                original_indices: original_indices.clone(),
+                                data: parallel_unit_data,
+                            }
```
[the remark this answers is not in the document]
Comment (shanicky) [reply src/meta/src/manager/catalog/fragment.rs:652]: ok, let me handle it

**Comment by shanicky** [comment]
I found a bug in the case of cascading updates, even if the upstream and downstream in the `StreamActor` of the newly created actor were already correct, the `post_apply_reschedules` process would still modify it, which causes duplicate actors to be added to upstream and downstream (removing is idempotent so there is no problem) and would affect the next reschedule, I skipped this part in the commit 1b78ea3595ee1f6b7f1d7f0dd5ea6cd6afebc47f and added some assert

File: src/meta/Cargo.toml [hunk 7]
```diff
@@ -23,6 +22,8 @@ futures = { version = "0.3", default-features = false, features = ["alloc"] }
 hex = "0.4"
 hyper = "0.14"
 itertools = "0.10"
+lazy_static = "1"
```
Comment (BugenZhao) [inline src/meta/Cargo.toml:25]:
```suggestion
```

File: src/meta/src/stream/stream_manager.rs [hunk 8]
```diff
@@ -745,6 +736,28 @@ where
         Ok(())
     }
 }
+
+fn vnode_mapping_to_actor_mapping(
```
Comment (BugenZhao) [inline src/meta/src/stream/stream_manager.rs:740]: Let's put this function along with `actor_to_vnode` and `actor_mapping_from_bitmaps`?

Comment (shanicky) [reply src/meta/src/stream/stream_manager.rs]: moved

File: src/meta/src/manager/catalog/fragment.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/meta/src/manager/catalog/fragment.rs:635]: Great. I think it's okay to have some minor behavior differences for reusing the scaling branch for the migration.

**Comment by BugenZhao** [comment]
> I found a bug in the case of cascading updates, even if the upstream and downstream in the `StreamActor` of the newly created actor were already correct, the `post_apply_reschedules` process would still modify it, which causes duplicate actors to be added to upstream and downstream (removing is idempotent so there is no problem) and would affect the next reschedule, I skipped this part and added some assert

I'm considering whether we can add some debug checks for this, like the validity of the meta store or the consistency between the meta's stream graph and the actual graph on compute nodes. 🤔

**Review by yezizp2012 — APPROVED** [review]

**Review by BugenZhao — APPROVED** [review]
LGTM!!! I've slightly tested this PR and it looks to work well. 🥰 We may gradually refine the code structure (like migration branch) and fix some corner cases (like no-shuffle #5423) in the future.

**Comment by shanicky** [comment]
> LGTM!!! I've slightly tested this PR and it looks to work well. 🥰 We may gradually refine the code structure (like migration branch) and fix some corner cases (like no-shuffle #5423) in the future.

Maybe we can start by adding some unit tests to the `post_apply_reschedules`, let me figure out how to do that, and I'll do it in the next PR later
